File: KthToLastNode/ktl.js
// Given the head and number signifying how many places from end

function kthToLastNode(k, head) {
    // Set up error handling
    if (k < 1) {
        throw new Error('Cannot find less than first to last node, ', k);
    }

    // Make two pointers - right traverses whole list, left one behind (k)
    let leftNode = head;
    let rightNode = head;

    // Initialize loop to move the right node
    for (let i = 0; i < k - 1; i++) {
        // If no next, k is larger than the list
        if (!rightNode.next) {
            throw new Error('k is larger than the list length');
        }
        rightNode = rightNode.next;
    }

    // Walk down the whole list
    while (rightNode.next) {
        leftNode = leftNode.next;
        rightNode = rightNode.next;
    }

    // When while stops being true, left is kth to last node
    return leftNode.value;
}

class ListNode {
    constructor(value) {
      this.value = value;
      this.next = null;
    }
  }
  
  let a = new ListNode("Australian Sheperd");
  let b = new ListNode("Beagle");
  let c = new ListNode("Cairne Terrier");
  let d = new ListNode("Dobermann");
  let e = new ListNode("English Mastiff");
  
  a.next = b;
  b.next = c;
  c.next = d;
  d.next = e;
  
  /* Some console.log tests */
  console.log(kthToLastNode(2, a));   // should print 'Dobermann'
  console.log(kthToLastNode(5, a));   // should print 'Australian Sheperd'
  console.log(kthToLastNode(3, c));   // should print 'Cairne Terrier'

File: KthToLastNode/ktl.py
# Use two pointers - 1 to reach end, 1 to signify k-from-end

class ListNode:
    def __init__(self, value):
        self.value = value
        self.next = None

def kthToLastNode(k, head): 
    # Handle errors first 
    if k < 1:
        raise Exception('Cannot go lower than 1st from last node')

    # Set up the pointers we need and attach to head
    leftNode = head
    rightNode = head

    # Set up for loop for right node (confirms it has a next)
    for i in range(k - 1):
        # If there isn't a next, error 
        if not rightNode.next:
            raise Exception('K is bigger than the linked list')
        rightNode = rightNode.next

    # Set up while loop for left/right (if so, increments both)
    while rightNode.next:
        leftNode = leftNode.next
        rightNode = rightNode.next

    return leftNode.value

a = ListNode("Australian Sheperd")
b = ListNode("Beagle")
c = ListNode("Cairne Terrier")
d = ListNode("Dobermann")
e = ListNode("English Mastiff")

a.next = b
b.next = c
c.next = d
d.next = e

# Some tests
print(kthToLastNode(2, a))   # should print 'Dobermann'
print(kthToLastNode(5, a))   # should print 'Australian Sheperd'
print(kthToLastNode(3, c))   # should print 'Cairne Terrier'
